projects/rocprofiler-sdk/source/lib/rocprofiler-sdk-tool/tool.cpp

```diff
@@ -2840,7 +2844,18 @@ tool_detach(void* /*tool_data*/)
     if(tool_metadata->process_end_ns == 0)
         rocprofiler_get_timestamp(&(tool_metadata->process_end_ns));
 
-    generate_output(cleanup_mode::reset);
+    // Launch generate output in a separate thread for safe cleanup
+    // This allows tool_detach to return immediately without waiting
+    {
+        std::lock_guard<std::mutex> lock(::cleanup_thread_mutex);
+        if(::cleanup_thread)
+        {
+            if(::cleanup_thread->joinable()) ::cleanup_thread->join();
+            ::cleanup_thread.reset();
+        }
+        ::cleanup_thread =
+            std::make_unique<std::thread>([]() { generate_output(cleanup_mode::reset); });
+    }
```

Comment:
```suggestion
        if(!::cleanup_thread || !::cleanup_thread->joinable())
        {
            ::cleanup_thread =
                std::make_unique<std::thread>([]() { generate_output(cleanup_mode::reset); });
        }
    }
```

Another room for improvement is to skip `generate_output` when another active thread is running, since the active thread will handle these results.

Reply:
> Another room for improvement is to skip `generate_output` when another active thread is running, since the active thread will handle these results.

This would be quite dangerous. We shouldn't have a thread generating output while data is being collected, our data structures at present could not handle that.